addon/adapter.js

```diff
@@ -261,45 +262,39 @@ export default DS.Adapter.extend({
 
     return new Ember.RSVP.Promise(function(resolve, reject) {
       options.success = function(payload, textStatus, jqXHR) {
-        let response;
-
-        if (!(response instanceof DS.AdapterError)) {
-          response = adapter.handleResponse(
-            jqXHR.status,
-            parseResponseHeaders(jqXHR.getAllResponseHeaders()),
-            response || payload,
-            options
-          );
-        }
-
-        if (response instanceof DS.AdapterError) {
-          Ember.run(null, reject, response);
+        let response = adapter.handleResponse(
+          jqXHR.status,
+          parseResponseHeaders(jqXHR.getAllResponseHeaders()),
+          payload,
+          options
+        );
+
+        if (response && response.isAdapterError) {
+          Ember.run.join(null, reject, response);
         } else {
-          Ember.run(null, resolve, response);
+          Ember.run.join(null, resolve, response);
```

Comment:
That's what the current implementation of the error handler in the ember-data rest serializer :)